Add ShuffleStage HLG Layer

rjzamora opened the pull request:
Depends on #6510 

Implements `ShuffleStage` - A simple `HighLevelGraph` (HLG) `Layer` for a single stage of a task-based shuffle in `dask.dataframe`.  To enable culling (without the need to materialize the full graph), the shuffling algorithm was revised to only include tasks that are needed to produce a specific set of output keys.

**TODO**:
- [x] Validate behavior and add tests
- [x] Benchmark the new approach

cc @madsbk @quasiben @mrocklin

mrocklin commented on dask/dataframe/shuffle.py:
```diff
@@ -24,6 +25,126 @@
+    def __getitem__(self, key):
+        return self._dict[key]
+
+    def __iter__(self):
+        return iter(self._dict)
+
+    def __len__(self):
+        return len(self._dict)
```
Do you have any thoughts on how easy it would be to generate `keys` and `get_dependencies` without generating the full graph?  These are likely to be the next useful steps here in terms of improving performance.

rjzamora replied:
Sorry for the delayed response here!  I'm not sure I fully understand the question,  because I still don't really understand when `keys` (which I assume is gerated with `__iter__`) and `get_dependencies` are actually *used* (or needed).

For example, it doesn't seem to me like there is ever a need to call `get_dependencies` on a `ShuffleStage` Layer, so I'm not really sure how to explore that idea.  Also, it is my understanding that the "dunder" methods you highlighted are only needed to operate on the layer after it already needs to be converted to a dictionary.   I am well aware that I have gaps in my understanding here, so feel free to set me straight :)

rjzamora commented:
**Update**:  I added a simple test and did some simple benchmarking.  Overall, performance is looking good:


```python
from dask.datasets import timeseries
from dask.dataframe.optimize import optimize
from dask.dataframe.shuffle import shuffle

ddf_d = timeseries(start='2000-01-01', end='2015-01-01', partition_freq='1d')
%timeit shuffle(ddf_d, "id", shuffle="tasks")
```
The `shuffle` API call is now very fast, because we are not actually constructing the graph yet...

**New**: `8.25 ms ± 15.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)`

**Old**: `414 ms ± 3.07 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)`

```python
dict_size = len(dict(shuffle(ddf_d, "id", shuffle="tasks").dask))
dict_size
```
The initial (pre-culling) graph size is now a bit smaller (when the graph is materialized), because the new algorithm is slightly more efficient...

**OLD**: `367063`

**NEW**: `355399`

```python
ddf = shuffle(ddf_d, "id", shuffle="tasks")
keys = [(ddf._name, i) for i in range(ddf.npartitions)]
%timeit optimize(ddf.dask, keys)
```
The optimization pass is now significantly faster...

**OLD**: `6.75 s ± 21.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)`

**NEW**: `2.72 s ± 30.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)`

Note that the filal (culled) graph is also a bit smaller than before (`348692` vs `354524`).

rjzamora commented:
cc @madsbk (Thanks for laying out all the HLG/Layer groundwork :) - It would be good to get your eyes on this)

madsbk commented on dask/dataframe/tests/test_shuffle.py:
```diff
@@ -1007,3 +1009,30 @@ def test_set_index_overlap():
+
+    # Since we already culled the HLG,
+    # culling the dictionary should not change the graph
+    dsk_dict = ensure_dict(dsk)
```
Let's work on eliminating `ensure_dict()`

```suggestion
    dsk_dict = dict(dsk)
```

madsbk commented on dask/dataframe/tests/test_shuffle.py:
```diff
@@ -17,6 +17,8 @@
 from dask.dataframe._compat import tm, assert_categorical_equal
 from dask import delayed
 from dask.base import compute_as_if_collection
+from dask.utils import ensure_dict
```
```suggestion
```

madsbk reviewed (approved):
Nice work @rjzamora, I only have some minor suggestions

madsbk commented on dask/dataframe/shuffle.py:
```diff
@@ -24,6 +25,126 @@
+            self.name, self.stage, self.k, self.npartitions
+        )
+
+    @property
+    def _dict(self):
```
```suggestion
    def materialize(self):
```

jakirkham commented:
@TomAugspurger, do you have any thoughts on this? 🙂

mrocklin reviewed (commented):
Oops.  I had a bunch of comments queued. My apologies for the delay.

> The optimization pass is now significantly faster...

Note that we're caching things, so the `timeit` call here may not be representative

rjzamora commented on dask/dataframe/shuffle.py:
```diff
@@ -24,6 +25,126 @@
+
+    def __init__(
+        self,
+        ddf,
```
Good call- Thanks for pointing this out!

rjzamora commented on dask/dataframe/shuffle.py:
```diff
@@ -24,6 +25,126 @@
+                self.n,
+                self.k,
+                self.ignore_index,
+            )
```
Right - Make sense to me

rjzamora commented:
Thanks for the review @mrocklin ! Hopefully I addressed your comments.

> Note that we're caching things, so the timeit call here may not be representative

Good point - I updated the "benchmark" results [above](https://github.com/dask/dask/pull/6650#issuecomment-700208535).  We still get a nice performance bump (albeit a slightly smaller one).

TomAugspurger commented on dask/dataframe/shuffle.py:
```diff
@@ -532,6 +723,78 @@ def _simple_rearrange_by_column_tasks(df, column, npartitions, ignore_index=Fals
+    return df2
+
+
+def rearrange_by_column_tasks_original(
+    df, column, max_branch=32, npartitions=None, ignore_index=False
```
What's the plan with original implementation?  Can it be removed?

rjzamora replied:
Yup! Completely forgot that I left the original version there while implementing the new version :)

TomAugspurger commented on dask/dataframe/shuffle.py:
```diff
@@ -24,6 +25,196 @@
+    def _dict(self):
+        """Materialize full dict representation"""
+        if hasattr(self, "_cached_dict"):
+            return self._cached_dict["dsk"]
```
What's the reason for `_cached_dict`nesting the Dask dict under the `"dsk"` key, rather than just returning the `dsk` itself?

rjzamora replied:
No reason at all - Another detail I forgot to clean up.  The very first version of the shuffle layer was doing something similar to Blockwise, and so we were caching a BasicLayer object as well as `dsk`.

TomAugspurger commented on dask/dataframe/shuffle.py:
```diff
@@ -24,6 +25,196 @@
+        inputs,
+        stage,
+        npartitions,
+        n,
+        k,
```
descriptive parameter names here would be nice. Suggestions: `n_input_partitions`, and `n_groups_per_partition`

rjzamora replied:
Right - We should definitely adopt better names. How about `n` -> `npartitions_input` and `k` -> `nsplits`?

TomAugspurger commented on dask/dataframe/shuffle.py:
```diff
@@ -24,6 +25,196 @@
+        self.n = n
+        self.k = k
+        self.ignore_index = ignore_index
+        self._name = dep_name
+        self._meta = dep_meta
```
Can these be called `_dep_name` and `_dep_meta`? Any reason to make them private while the others are public?

rjzamora replied:
No good reason - I'll change to `self.dep_name` and `self.dep_meta`

TomAugspurger commented on dask/dataframe/shuffle.py:
```diff
@@ -24,6 +25,196 @@
+        for key in keys:
+            try:
+                _name, _part = key
+            except ValueError:
+                continue
```
I'm curious about when this occurs. We're saying that `keys` might be things like `(name, int)` or just something like `other_name`. Do you have a sense for which is more common? Catching exceptions can be relatively expensive if they happen often.

But perhaps this isn't the performance bottleneck and can be ignored.

rjzamora replied:
> I'm curious about when this occurs. We're saying that keys might be things like (name, int) or just something like other_name. Do you have a sense for which is more common? Catching exceptions can be relatively expensive if they happen often.

I don't believe this exception is very unlikely to happen (it never happens in any tests).  The purpose of this method is to extract a list of partition IDs (integers) from the keys being used to cull the `ShuffleStage` layer.  The input list is currently guaranteed to contain keys that look like: `(self.name, int)`.  Therefore, the `try/except` and `_name != self.name` logic is not expected to be used in practice (its only there to "harden" the method a bit)... Let me know if you'd like to be more explicit and just raise an error if the key is something unexpected.

TomAugspurger commented on dask/dataframe/shuffle.py:
```diff
@@ -24,6 +25,196 @@
+            self.ignore_index,
+            self._name,
+            self._meta,
+            parts_out=self._keys_to_parts(keys),
+        )
+        culled_deps = self._cull_dependencies(keys)
```
I see that `_cull_dependencies` calls `_keys_to_parts`. If that's expensive, then `cull_dependencies` could take an optional `parts_out` argument. Then we compute it once here and pass it to `cull_dependencies`.

rjzamora replied:
Good suggestion - I'll make the change.

rjzamora commented on dask/dataframe/shuffle.py:
```diff
@@ -24,6 +25,196 @@
 logger = logging.getLogger(__name__)
 
 
+class ShuffleStage(Layer):
```
I'm open to `ShuffleLayer`.  However, I went with the `ShuffleStage` name, becuase the layer does not correspond to a full shuffle operation in many cases (just a single stage).  Perhaps that is not an issue?

rjzamora replied:
Okay - After a bit of thought, I decided the `ShuffleLayer` name is probably best. I made the change :)

rjzamora commented:
**Update**:  I realized this morning that this PR was leaving out the common case of the "simple" shuffle (where the number of output partitions is less than `max_branch`).  I revised the class structure to include both a `SimpleShuffleLayer` and a `shuffleLayer` (which inherits from `SimpleShuffleLayer`).

rjzamora commented on dask/dataframe/shuffle.py:
```diff
@@ -24,6 +26,296 @@
+        all input partitions. This method does not require graph
+        materialization.
+        """
+        deps = defaultdict(set)
```
@madsbk - I just wanted to get your thoughts on something here...

It seems that the HLG `cull` operation breaks when this `deps` dictionary is a vanilla `dict` (rather than a `defaultdict(set)`).  The problem is that we are not actually returning any intra-layer dependencies when the shuffle layer is culled (because we are not actually materializing the graph). This means, we will get a keyerror during this [culled_deps[k]](https://github.com/dask/dask/blob/51b76e5d4820c435c4ee179ef31ea9916ec229c8/dask/highlevelgraph.py#L435) access in `HighLevelGraph.cull` (because we are not adding dependencies for **all** keys).

Does this seem like it could be a problem?

madsbk replied:
This should be fixed in https://github.com/dask/dask/pull/6699, in which `cull()` is only required to return _external_ dependencies.

rjzamora replied:
Thanks @madsbk !

mrocklin commented:
@jrbourbeau I'd like to put this on your queue as well

jrbourbeau commented on dask/dataframe/tests/test_shuffle.py:
```diff
@@ -1007,3 +1008,30 @@ def test_set_index_overlap():
+
+    # Since we already culled the HLG,
+    # culling the dictionary should not change the graph
+    dsk_dict = dict(dsk)
+    dsk_dict_culled, _ = cull(dsk_dict, keys)
+    assert len(dsk_dict_culled) < len(dsk_dict)
```
My understanding about what we're trying to test here could be wrong, but based on the comment above I would have expected us to cull the already culled high-level graph, `dsk_culled`, and then assert that the the graph is the same before and after culling (since it has already been culled earlier):

```suggestion
    dsk_dict = dict(dsk_culled)
    dsk_dict_culled, _ = cull(dsk_dict, keys)
    assert dsk_dict_culled == dsk_dict
```

rjzamora replied:
good catch!  My mistake.

jrbourbeau reviewed (commented):
Thanks for all your work here @rjzamora!

Please correct me if I'm wrong, but from what I can tell this PR moves existing DataFrame shuffling code into two new `SimpleShuffleLayer` and `ShuffleLayer` layer classes. The corresponding low-level task graphs aren't materialized until we try to inspect the underlying task graph (e.g. `__getitem__`) which will eventually help us send a smaller object to the scheduler (once we're directly sending `HighLevelGraphs`).

One piece of follow-up work that comes to mind is adding a custom serialization method for these new shuffle layers. The implementation over in https://github.com/dask/dask/pull/6693 would cause us to fully materialize the task graph for shuffle layers before sending them to the scheduler. Though that's certainly future work we don't need to worry about for this PR : )  

Overall I think the changes here look good. The fact that tests are passing give me confidence. I've left a few small comments, but otherwise this appears to be good to go. Would you recommend we merge this in?

rjzamora commented on dask/dataframe/tests/test_shuffle.py:
```diff
@@ -1007,3 +1008,30 @@ def test_set_index_overlap():
+
+    # Check HLG layers beginning with "shuffle-"
+    for name, layer in dsk.layers.items():
+        if name.startswith("shuffle-"):
+            assert isinstance(layer, dd.shuffle.ShuffleLayer)
```
I added this check to make sure we are actually using `ShuffleLayer` layers.  So, I worry that if we aren't adding them, then this change will miss it?

jrbourbeau replied:
Gotcha! I has the concern that if we change our `ShuffleLayer` naming scheme then we would start to miss this check. I just pushed a small commit which `assert`s that there are `ShuffleLayer`s in the HLG and that the names of the `ShuffleLayer`s are as expected (i.e. they start with `"shuffle-"`). That should, I think, handle both of our concerns

rjzamora commented on dask/dataframe/shuffle.py:
```diff
@@ -24,6 +25,320 @@
+    ignore_index: bool, default False
+        Ignore index during shuffle.  If ``True``, performance may improve,
+        but index values will not be preserved.
+    dep_name : str
+        Name of input collection.
+    dep_meta : pd.DataFrame-like object
+        Empty metadata of input collection.
```
That seems fine to me. I had a lot of trouble deciding on the best names, so I eventually picked something arbitrary and moved on - Thank you for being decisive for me :)

jrbourbeau reviewed (approved):
Thanks @rjzamora! Will merge once CI passes

rjzamora commented:
Thanks for your help @jrbourbeau !

Sounds good.  My only thought: I'm not sure I understand the purpose of `f5dfd60`, since we will no longer catch the case that the `ShuffleLayer` layers are missing.

jrbourbeau commented:
The

```python
    assert any(
        isinstance(layer, dd.shuffle.ShuffleLayer) for layer in dsk.layers.values()
    )
```

check should ensure that `ShuffleLayer` layers aren't missing

rjzamora commented:
Ah - sorry!  I was looking at the wrong change and didn't see that.  In that case, I don't think the `assert name.startswith("shuffle-")` check is really necessary (since we "could" use a different name), but it doesn't doesn't... So, I think we are good.  Thanks again!
